packages/client/src/authentication/Authentication.ts

```diff
@@ -1,39 +1,61 @@
-import { GraphQLClient } from 'graphql-request';
+import { failure, PromiseResult, success } from '@lens-protocol/shared-kernel';
 
 import { LensConfig } from '../consts/config';
-import { getSdk, Sdk } from './graphql/auth.generated';
-
-export type Credentials = { accessToken: string; refreshToken: string };
+import { CredentialsExpiredError, NotAuthenticatedError } from '../consts/errors';
+import { AuthenticationApi } from './adapters/AuthenticationApi';
+import { CredentialsStorage } from './adapters/CredentialsStorage';
+import { InMemoryStorageProvider } from './adapters/InMemoryStorageProvider';
 
 export class Authentication {
-  private readonly sdk: Sdk;
+  private readonly api: AuthenticationApi;
+  private readonly storage: CredentialsStorage;
 
   constructor(config: LensConfig) {
-    const client = new GraphQLClient(config.environment.gqlEndpoint);
-    this.sdk = getSdk(client);
+    this.api = new AuthenticationApi(config);
+    this.storage = new CredentialsStorage(config.storage || new InMemoryStorageProvider());
   }
 
-  async challenge(address: string): Promise<string> {
-    const result = await this.sdk.AuthChallenge({ address });
-
-    return result.data.result.text;
+  async generateChallenge(address: string): Promise<string> {
+    return this.api.challenge(address);
   }
 
-  async verify(accessToken: string): Promise<boolean> {
-    const result = await this.sdk.AuthVerify({ accessToken });
-
-    return result.data.result;
+  async authenticate(address: string, signature: string): Promise<void> {
+    const credentials = await this.api.authenticate(address, signature);
+    await this.storage.set(credentials);
   }
 
-  async authenticate(address: string, signature: string): Promise<Credentials> {
-    const result = await this.sdk.AuthAuthenticate({ address, signature });
-
-    return result.data.result;
+  async isAuthenticated(): Promise<boolean> {
+    const credentials = await this.storage.get();
+    return credentials ? !credentials.isExpired() : false;
   }
 
-  async refresh(refreshToken: string): Promise<Credentials> {
-    const result = await this.sdk.AuthRefresh({ refreshToken });
+  async getRequestHeader(): PromiseResult<
```

Comment:
this is the core of this PR, a single method, called before each request that requires auth, see Reactions.ts how it is used. It is a lazy way of checking expiration or refreshing token. And will call auth API only if required.

Reply:
At the moment `Authentication` is an internal of the client SDK. If the plane of exposing individual submodules to very advanced users is still in place, this method should be segregated behind a privileged interface that only submodules can read.

Reply:
yes, I want to keep the option to use submodules independently open. I'm using that in submodule tests.
how would the privileged interface look like (is it just ts definition?) and would it be possible to access the `getRequestHeader` method in tests?

Reply:
Let's cross that bridge once we are in a position to export `Authentication`. Can you please keep track of this aspect so we can have that conversation later?

Reply:
task created for the `IAuthentication` public interface https://avara.height.app/lens-sdk/T-6208